// docpad/out/main.js
// Generated by CoffeeScript 1.9.1
(function() {
  var Backbone, DocPad, createInstance, createMiddlewareInstance, pathUtil, queryEngine, ref;

  pathUtil = require('path');

  ref = require('./lib/docpad'), DocPad = ref.DocPad, queryEngine = ref.queryEngine, Backbone = ref.Backbone, createInstance = ref.createInstance, createMiddlewareInstance = ref.createMiddlewareInstance;

  module.exports = {
    DocPad: DocPad,
    queryEngine: queryEngine,
    Backbone: Backbone,
    createInstance: createInstance,
    createMiddlewareInstance: createMiddlewareInstance,
    require: function(relativePath) {
      var absolutePath, err;
      absolutePath = pathUtil.normalize(pathUtil.join(__dirname, relativePath));
      if (absolutePath.replace(__dirname, '') === absolutePath) {
        throw new Error("docpad.require is limited to local docpad files only: " + relativePath);
      }
      try {
        require.resolve(absolutePath);
      } catch (_error) {
        err = _error;
        absolutePath = pathUtil.join(__dirname, 'lib', relativePath);
        require.resolve(absolutePath);
      }
      return require(absolutePath);
    }
  };

}).call(this);
